torch/csrc/jit/passes/fixup_trace_scope_blocks.cpp

```diff
@@ -0,0 +1,350 @@
+        if (i < prefix.size()) {
+          TORCH_INTERNAL_ASSERT(sub_atoms[i] == prefix[i]);
+        } else {
+          callee_val = g->insertGetAttr(callee_val, sub_atoms[i]);
+          callee_mod = callee_mod.get_module(sub_atoms[i]);
+        }
+      }
+
+      createMethodCalls(n->g(attr::Subgraph), callee_mod, sub_atoms);
+
+      Function* f = nullptr;
+
+      // Find a unique name to add this method as
+      // We try forward, forward1, forward2, ...
+      for (size_t method_idx = 0;; method_idx++) {
+        std::string method_name = "forward";
+        if (method_idx != 0) {
+          method_name += std::to_string(method_idx);
+        }
+        if (callee_mod.find_method(method_name)) {
+          continue;
+        } else {
+          auto qualname = c10::QualifiedName(callee_mod.name(), method_name);
+          f = callee_mod.class_compilation_unit()->create_function(
+              qualname, n->g(attr::Subgraph));
+          callee_mod.type()->addMethod(f);
+          break;
+        }
+      }
+      TORCH_INTERNAL_ASSERT(f);
+      std::vector<NamedValue> nvs = {
+          NamedValue(callee_val->node()->sourceRange(), callee_val)};
+      for (Value* i : n->inputs()) {
+        nvs.emplace_back(i->node()->sourceRange(), i);
+      }
+      auto schema =
+          script::matchSchema(f->getSchema(), n->sourceRange(), *g, nvs, {});
+      Value* retval = g->insertMethodCall(f->qualname().name(), schema);
+      n->output()->replaceAllUsesWith(retval);
+      n->destroy();
+    }
+  }
+}
+
+void inlineScopeBlocks(Block* b) {
+  for (auto n_itr = b->nodes().begin(); n_itr != b->nodes().end();) {
+    Node* n = *n_itr++;
+    for (Block* sub_b : n->blocks()) {
+      inlineScopeBlocks(sub_b);
+    }
+    if (n->kind() == prim::FakeScopeBlock) {
+      // Convert the block to a graph so we can inline it
+      auto graph = std::make_shared<Graph>();
+      std::unordered_map<Value*, Value*> remaps;
+      graph->block()->cloneFrom(n->blocks()[0], [&](Value* v) {
+        remaps[v] = graph->block()->addInput()->copyMetadata(v);
+        n->addInput(v);
+        return remaps[v];
+      });
+
+      inlineCallTo(n, *graph);
+    }
+  }
+}
+
+void runCleanupPasses(const std::shared_ptr<Graph>& g) {
+  if (script::getInlineEverythingMode()) {
+    Inline(*g);
+  }
+  LowerSimpleTuples(g);
+  EliminateDeadCode(g);
+  LintGraph(g);
+}
+
+void runCleanupPasses(script::Module* m) {
+  auto methods = m->get_methods();
+  for (auto module : m->get_modules()) {
+    runCleanupPasses(&module.module);
+  }
+  for (auto& method : methods) {
+    runCleanupPasses(method.graph());
+  }
+}
+
+void FixupTraceScopeBlocks(
+    std::shared_ptr<Graph>& graph,
+    script::Module* self) {
+  // Gather level information about blocks in the graph.
+  // Assign each block a number, with graph->block() gettng
+  // 0 and each sub-block gets a higher number based on how
+  // deeply nested it is.
+  BlockInfoMap block_info;
```

Comment:
We should put level information into the Block itself. It does not change once a block is created so it is just `owningNode()->block()->level() + 1`